src/prob/expansion.jl

```diff
@@ -0,0 +1,493 @@
+#### General Assumptions of these Expansion Models ####
+#
+# - if the branch has a field called "construction_cost", then it is assumed to be a buildable edge
+#
+
+export run_expansion
+
+function run_expansion(file, model_constructor, solver; kwargs...)
+    return run_generic_model(file, model_constructor, solver, post_expansion; solution_builder = get_expansion_solution, kwargs...) 
+end
+
+# the general form of the expansion optimization model
+function post_expansion{T}(pm::GenericPowerModel{T})
+    build_expansion_sets(pm) ## create the data sets we need
+    
+    variable_line_expansion(pm) 
+    variable_complex_voltage_on_off(pm)
+
+    variable_active_generation(pm)
+    variable_reactive_generation(pm)
+
+    variable_active_line_flow(pm)
+    variable_reactive_line_flow(pm)
+
+    objective_expansion_cost(pm)
+
+    constraint_theta_ref(pm)
+    constraint_complex_voltage_expansion_on_off(pm)
+
+    for (i,bus) in pm.set.buses
+        constraint_active_kcl_shunt(pm, bus)
+        constraint_reactive_kcl_shunt(pm, bus)
+    end
+
+    for i in pm.data["new_branches"]
+        branch = pm.set.branches[i] 
+        constraint_active_ohms_yt_expansion_on_off(pm, branch)
+        constraint_reactive_ohms_yt_expansion_on_off(pm, branch) 
+
+        constraint_phase_angle_difference_expansion_on_off(pm, branch)
+
+        constraint_thermal_limit_from_expansion_on_off(pm, branch)
+        constraint_thermal_limit_to_expansion_on_off(pm, branch)
+    end
+    
+    for (i,branch) in pm.set.branches
+        if !in(i,pm.data["new_branches"])    
+            constraint_active_ohms_yt_expansion(pm, branch)
+            constraint_reactive_ohms_yt_expansion(pm, branch)
+            constraint_phase_angle_difference_expansion(pm, branch)
+            constraint_thermal_limit_from(pm, branch)
+            constraint_thermal_limit_to(pm, branch)
+        end

```

Comment:
On 1) Yes, I was not aiming to be critical, but more wanted to start brainstorming how to handle that set of constraints.  It was the place I felt the least comfortable when implementing the expansion variables.

On 2) Yes, Matpower does not support EP.  If you pulled in a matpower case, this model would assume there is no expansion options unless a supplimental data file was provided.

On the comment "In these cases, why do you need to make new _expansion variants of the constraints, why not simply use the built in ones?".  Yes, I thought about this, but this would preclude creating a problem with expansion and switching options... one way to do this would be to split each of the on/off constraints into two functions.  One function takes as an argument the on/off variable and constructs the on/off constraint with this variable.  The second function(s) declares we will use the switching variable or the expansion variable and then calls the first function....